File: coinChange.py
# Time Complexity : O(m*n)
# Space Complexity : O(m*n)

class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        row = len(coins) + 1
        col = amount +1

        dp = [[99999 for x in range(col)] for x in range(row)]

        for j in range (1, row):
            dp[j][0] = 0

        for i in range(1,row):
            for j in range(1,col):
                if coins[i-1]>j:
                    dp[i][j]=dp[i-1][j]
                else:
                    dp[i][j]=min(dp[i-1][j],1+dp[i][j-coins[i-1]])
        
        if (dp[row-1][col-1] == 99999):
            return -1
        return dp[row-1][col-1]

File: coin change.cpp

// This is an exhaustive approach

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        
        return helper(coins,0,amount,0);
    }
    
    int helper(vector<int>& coins,int index, int amount, int min)
    {
        
        //base case:
        //valid path
        if(amount ==0)
        {
            return min;
        }
        //invalid path out of bound
        if(amount<0 or index==coins.size())
        {
            return -1;
        }
        
        
        //case 1: choose:
        int case1= helper(coins, index, amount-coins[index], min+1);
        
        
        //case2: not choose:
        int case2= helper(coins, index+1, amount, min);
        
        
        if(case1==-1)return case2;
        if(case2==-1)return case1;
        
        return std::min(case1,case2);
    }
};

